dispatcher/brokers/pg_notify.py

```diff
@@ -13,87 +16,108 @@
-#     "default": {
-#         "ENGINE": "django.db.backends.postgresql",
-#         "HOST": os.getenv("DB_HOST", "127.0.0.1"),
-#         "PORT": os.getenv("DB_PORT", 55777),
-#         "USER": os.getenv("DB_USER", "dispatch"),
-#         "PASSWORD": os.getenv("DB_PASSWORD", "dispatching"),
-#         "NAME": os.getenv("DB_NAME", "dispatch_db"),
-#     }
-# }
-
-
-async def aget_connection(config):
-    return await psycopg.AsyncConnection.connect(**config, autocommit=True)
-
-
-def get_connection(config):
-    return psycopg.Connection.connect(**config, autocommit=True)
-
-
-async def aprocess_notify(connection, channels, connected_callback=None):
-    async with connection.cursor() as cur:
-        for channel in channels:
-            await cur.execute(f"LISTEN {channel};")
-            logger.info(f"Set up pg_notify listening on channel '{channel}'")
-
-        if connected_callback:
-            await connected_callback()
-
-        while True:
-            logger.debug('Starting listening for pg_notify notifications')
-            async for notify in connection.notifies():
-                yield notify.channel, notify.payload
-
-
-async def apublish_message(connection, channel, payload=None):
-    async with connection.cursor() as cur:
-        if not payload:
-            await cur.execute(f'NOTIFY {channel};')
-        else:
-            await cur.execute(f"NOTIFY {channel}, '{payload}';")
-
-
-def get_django_connection():
-    try:
-        from django.conf import ImproperlyConfigured
-        from django.db import connection as pg_connection
-    except ImportError:
-        return None
-    else:
-        try:
-            if pg_connection.connection is None:
-                pg_connection.connect()
-            if pg_connection.connection is None:
-                raise RuntimeError('Unexpectedly could not connect to postgres for pg_notify actions')
-            return pg_connection.connection
-        except ImproperlyConfigured:
-            return None
-
-
-def publish_message(queue, message, config=None, connection=None, new_connection=False):
-    conn = None
-    if connection:
-        conn = connection
-
-    if (not conn) and (not new_connection):
-        conn = get_django_connection()
-
-    created_new_conn = False
-    if not conn:
-        if config is None:
-            raise RuntimeError('Could not use Django connection, and no postgres config supplied')
-        conn = get_connection(config)
-        created_new_conn = True
-
-    with conn.cursor() as cur:
-        cur.execute('SELECT pg_notify(%s, %s);', (queue, message))
-
-    logger.debug(f'Sent pg_notify message to {queue}')
-
-    if created_new_conn:
-        conn.close()
+class PGNotifyBase(BaseBroker):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        if self._config:
+            self._config = self._config.copy()
+            self._config['autocommit'] = True
+
+
+class AsyncBroker(PGNotifyBase):
```

Comment:
Probably I missing something but I see some code duplication when handling the connection between the async broker and the sync broker and I wonder if would make sense to add the logic in PGNotifyBase using polymorphism to delegate connection instantiation and implement `create_connection` as class method or is it something that you have discarded for some reason?

Reply:
We can't share that method b/c one version is `def` and the other version is `async def`. Good eye. I don't think there's a solution to consolidate thta.

Reply:
An async func can call a sync func. 
I mean something like this in PGNotifyBase
```
    async def get_async_connection(self) -> psycopg.AsyncConnection:
    def get_sync_connection(self) -> psycopg.Connection:
    def create_connection(self, factory_attr, create_method):
```
plus:
```
class AsyncBroker(PGNotifyBase):
    async def get_connection(self) -> psycopg.AsyncConnection:
        return await self.get_async_connection()

class SyncBroker(PGNotifyBase):
    def get_connection(self) -> psycopg.Connection:
        return self.get_sync_connection()

```

Reply:
I didn't believe you at first, but you were right. I wound up with something in this general spirit. The key thing I was missing was that the factory method had to be passed around.